workflow/snakemake_rules/select_strains.smk

```diff
@@ -0,0 +1,129 @@
+'''
+This file contains rules that select strains for a build, extracts the sequences
+and the meta data subset.
+
+input:
+ - "data/lineage/metadata.tsv"
+ - "data/lineage/{segment}.fasta"
+ - "data/{lineage}/{center}_{passage}_{assay}_titers.tsv"
+
+output:
+ - builds/{build_name}/strains.txt
+ - builds/{build_name}/metadata.tsv
+ - builds/{build_name}/titers.tsv
+ - builds/{build_name}/{segment}/sequences.fasta
+'''
+
+build_dir = config.get("build_dir", "builds")
+
+def get_titers_for_build(w):
+    return "data/{lineage}/{center}_{passage}_{assay}_titers.tsv".format(**config['builds'][w.build_name])
+
+rule titer_priorities:
+    input:
+        titers = get_titers_for_build,
+        metadata = lambda w: f"data/{config['builds'][w.build_name]['lineage']}/metadata.tsv"
+    output:
+        priorities = build_dir + "/{build_name}/titer_priorities.tsv"
+    run:
+        import pandas as pd
+        metadata = pd.read_csv(input.metadata, sep='\t')
+        titer_counts = {s:0 for s in metadata.strain}
+
+        titers = pd.read_csv(input.titers, sep='\t')
+        for ri, row in titers.iterrows():
+            test_strain = row.iloc[0]
+            if test_strain in titer_counts:
+                titer_counts[test_strain] += 1
+
+        with open(output.priorities, 'w') as fh:
+            for s,p in titer_counts.items():
+                fh.write(f"{s}\t{p}\n")
+
+def get_subsample_input(w):
+    files = {"metadata": f"data/{config['builds'][w.build_name]['lineage']}/metadata.tsv"}
+    if config['builds'][w.build_name]['subsamples'][w.subsample].get('priorities', '')=='titers':
+        files['titers']=build_dir + f"/{w.build_name}/titer_priorities.tsv"
+    return files
+
+rule subsample:
+    input:
+        unpack(get_subsample_input)
```

Comment:
Instead of unpacking output from a wildcard function like this, what if we allow users to define the “titer priorities” file as a subsample input in the config file (e.g., `priorities_file: "{build_name}/titer_priorities.tsv"`)? This would allow Snakemake to build the titer priorities on the fly, as needed, and also allow users to define their own priorities file. This would decouple the *titers* bit from the *priorities* bit.